opal/mca/btl/sm/btl_sm_module.c

```diff
@@ -178,6 +179,7 @@ static int sm_btl_first_time_init(mca_btl_sm_t *sm_btl, int n)
             return rc;
         }
     }
+    OBJ_CONSTRUCT(&sm_btl->super.btl_atomic_fallback_lock, opal_mutex_t);
```

Comment:
Is that a double-init with https://github.com/open-mpi/ompi/pull/8574/files#diff-600bdda1e90d484b3451565f25a31b08d7cae44139e005c485d4441f03749525R589 ?

Reply:
To clarify: I tried to understand the mutex construction/destruction and couldn't find a matching destructor call for this construction here

Reply:
This is a good catch, it might be. Will check

Reply:
I moved the lock to base, I think it makes more sense to have it there, I think the btls should all use the same lock to preserve atomicity. That look good? If so I can squish it down.

@bosilca ^^

Reply:
Shouldn't the lock be per window instead of global ?

Reply:
The answer is not simple, it depends on the memory model. For the two you just mentionned, SM and TCP, it would work because they both do atomic operations in the target memory. Take SM and IB and you would have a totally different story but a less clear outcome.

In any case, the global lock is providing a stronger consistency model than required by MPI, with certain performance implications as well. A lock per window (not per BTL) will at least alleviate the cost of fighting for locks in a highly contentious case.

Reply:
I agree that it should be per-window,  not per-btl.

Hacking at the code though, I'm not sure of the best way to make this work. The btl has no knowledge of windows, and that's where the actual accumulate operation happens on the target - and where the lock needs to happen. I suspect we would need a largish refactor to make this work, but perhaps not and I am missing something?

I'm happy to take a crack at implementing this, but at this point I'm not sure how to proceed. It doesn't look to me like the current window has its own btl module pointer, it shares it from a global btl pointer with other windows. I guess the easiest thing to do would be to give each window it's own copy, but I don't know if there's other ramifications to that.

That said, this is such a small edge case, I wonder if it's worth the effort. Users who mis-align their windows shouldn't expect great performance, anyway.

Reply:
@hjelmn and @devreal do you have any thoughts on this?

This locking issue seems to be a bit thorny.

Reply:
I'm OK with a global lock. People using misaligned atomics should have no expectation of performance whatsoever and ripping the module structure apart for it isn't worth it. We should push the alignment discussion within the MPI forum to eventually disallow it.

Reply:
Yeah, I am thinking along the same lines.

One way we can do this on a per-window basis would be to give the btl a callback function into the rdma layer with window info to look it up on the target side. 

If the guts is already present to do something like (I don't think it is, but not 100% sure) that it wouldn't be too bad. But if we need to rip things apart to get that working, it probably isn't worth it.